modules/stochastic_tools/src/samplers/ParallelMarkovChainMonteCarloBase.C

```diff
@@ -0,0 +1,220 @@
+{
+  for (unsigned int j = 0; j < _num_parallel_proposals; ++j)
+  {
+    for (unsigned int i = 0; i < _priors.size(); ++i)
+      _new_samples[j][i] = _priors[i]->quantile(getRand(seed_value));
+  }
+}
+
+void
+ParallelMarkovChainMonteCarloBase::sampleSetUp(const SampleMode /*mode*/)
+{
+  if (_step < 1 || _check_step == _step)
+    return;
+  _check_step = _step;
+
+  unsigned int seed_value = _step > 0 ? (_step - 1) : 0;
+
+  // Filling the new_samples vector of vectors with new proposal samples
+  proposeSamples(seed_value);
+
+  // Draw random numbers to facilitate decision making later on
+  for (unsigned int j = 0; j < _num_parallel_proposals; ++j)
+    _rnd_vec[j] = getRand(seed_value);
+}
+
+void
+ParallelMarkovChainMonteCarloBase::randomIndex(const unsigned int & ub,
+                                               const unsigned int & exclude,
+                                               const unsigned int & seed,
+                                               unsigned int & req_index)
+{
+  req_index = exclude;
+  while (req_index == exclude)
+    req_index = getRandl(seed, 0, ub);
+}
+
+void
+ParallelMarkovChainMonteCarloBase::randomIndex2(const unsigned int & ub,
+                                                const unsigned int & exclude,
+                                                const unsigned int & seed,
+                                                unsigned int & req_index1,
+                                                unsigned int & req_index2)
+{
+  randomIndex(ub, exclude, seed, req_index1);
+  req_index2 = req_index1;
+  while (req_index1 == req_index2)
+    randomIndex(ub, exclude, seed, req_index2);
+}
+
+void
+ParallelMarkovChainMonteCarloBase::combineWithConfg()
+{
+  unsigned int index1;
+  int index2 = -1;
+  std::vector<Real> tmp;
+  for (unsigned int i = 0; i < _num_parallel_proposals * _confg_values[0].size(); ++i)
+  {
+    index1 = i % _num_parallel_proposals;
+    if (index1 == 0)
+      ++index2;
+    tmp = _new_samples[index1];
+    for (unsigned int j = 0; j < _confg_values.size(); ++j)
+      tmp.push_back(_confg_values[j][index2]);
+    _new_samples_confg[i] = tmp;
+  }
+}
+
+const std::vector<Real> &
+ParallelMarkovChainMonteCarloBase::getRandomNumbers() const
+{
+  return _rnd_vec;
+}
+
+const std::vector<Real> &
+ParallelMarkovChainMonteCarloBase::getVarSamples() const
+{
+  return _new_var_samples;
+}
+
+const std::vector<const Distribution *>
+ParallelMarkovChainMonteCarloBase::getPriors() const
+{
+  return _priors;
+}
+
+const Distribution *
+ParallelMarkovChainMonteCarloBase::getVarPrior() const
+{
+  return _var_prior;
+}
+
+Real
+ParallelMarkovChainMonteCarloBase::computeSample(dof_id_type row_index, dof_id_type col_index)
+{
+  if (_step < 1)
+  {
+    for (unsigned int i = 0; i < _num_parallel_proposals; ++i)
+      _new_samples[i] = _initial_values;
+  }
+
+  // Combine the proposed samples with experimental configurations
+  combineWithConfg();
```

Comment:
Could this be `combineWithEperimentalConfig()`? These are pretty high-level functions so it is likely that users will look at them.